### DISCUSSION TITLE
Make "always include" more granular

### DISCUSSION THREAD

**Discussion by vesavlad** [tailflow/laravel-orion#298] [Ideas]
I have an api for providing localization information and shaping data. In this case I don't want to always include points for a shape for example when doing the regular "index" call but I want them to always be included when doing the "detail" call to get one shape.

Ex:
```/api/v1/shapes``` should not include the points cause this generates allot of data that not all is useful for the client
```/api/v1/shapes/1``` should include the points that are in the shape. 

Can I achieve the following use-case without ```?includes=points```? As mentioned I want them the always be included and not leave this client.

Plugin version: 2.1.3
Laravel version: 8.49.2

**Comment by alexzarbn** [comment]
Hi @vesavlad,

Perhaps just put a condition inside the `alwaysInclude` method?

```php
protected function alwaysIncludes(): array
{
    if (Route::current()->named('<name of the route that always includes relations>')) {
        return ['points'];
    }

    return [];
}
```

**Comment by ionut-tanasa** [comment]
Maybe `indexIncludes` and `showIncludes` could be useful?

**Comment by vesavlad** [comment]
well, would say that indeed would make sense for orion to have something like this available: 
```php
protected function alwaysIncludes(): array
{
    return [
          'points' => IncludesOption::showOnIndex | IncludesOption::showOnDetail,
          'other' => IncludesOption::showOnIndex ,
          'parent',
          'child' => IncludesOption::showOnDetail,
    ];
}
```

Where IncludesOptions:
```php
final class IncludesOptions
{
    const showOnIndex = 1;
    const showOnDetail = 2;
    const showOnUpdate = 4;
    const showOnDelete = 8;
}
```

Where in case no config is specified we the include will be available for both.
More than this, this option can be extendend to other CRUD operations leaving the versatility for the user where he wants the data to be shown.

This option can be used also for ```includes()``` method.

In this way:
 - the backwards compatibility is not broken
 - user can bennefit from this if the wants and where he wants.

**Comment by alexzarbn** [comment]
Hmm, I do see some benefits of making whitelisting more granular, but at the same time it feels like covering of edge cases 🤔

I will leave the issue open - lets see, if there are more developers that encounter this situation. In the meantime, I will check what are the possible solutions for this.
